### PR TITLE
ADD evalPriority feature

### PR DESCRIPTION
**Pull request by fgalan**
Issue #4566

- [x] Re-align .test
- [x] New test: evalPriority fails + CRUD
- [x] Fix FIXME PR marks
- [x] Doc
- [x] CNR

### CODE REVIEW

**Comment by mapedraza** [comment]
May a test should implement attributes ordered by priority in a different way than alphabetically

**Comment by fgalan** [comment]
> May a test should implement attributes ordered by priority in a different way than alphabetically

Done in [3eed5ca](https://github.com/telefonicaid/fiware-orion/pull/4590/commits/3eed5ca99f9d89ed4f40f4b3a6408d3e15cd3447)

File: src/lib/common/JsonHelper.cpp [hunk 1]
```diff
@@ -422,17 +424,13 @@ void JsonVectorHelper::addNull(void)
...
 {
-  // This check allows to call str() several times (needed when this is used in ExprContext)
-  if (!closed)
-  {
-    ss += ']';
-    closed = true;
-  }
+  ss += ']';
   return ss;
 }
```
Comment (fgalan) [inline src/lib/common/JsonHelper.cpp:436]: This was introduced in 4.0.0, but it seems is not actually needed.

Going back to 3.12.0 version of this code.

NTC (informative)

File: src/lib/ngsi/ContextAttribute.cpp [hunk 2]
```diff
@@ -1124,11 +1144,11 @@ std::string ContextAttribute::toJson(const std::vector<std::string>&  metadataFi
   filterAndOrderMetadata(metadataFilter, &orderedMetadata);
 
   //
-  // metadata (note that ngsi field in custom notifications doesn't include metadata)
+  // metadata (note that ngsi field in custom notifications avoids empety metadata array, i.e. "metadata": {})
```
Comment (fgalan) [inline src/lib/ngsi/ContextAttribute.cpp:1147]:
```suggestion
  // metadata (note that ngsi field in custom notifications avoids empty metadata array, i.e. "metadata": {})
```

File: src/lib/ngsi/ContextAttribute.cpp [hunk 3]
```diff
@@ -1124,11 +1144,11 @@ std::string ContextAttribute::toJson(const std::vector<std::string>&  metadataFi
...
   //
-  if (!renderNgsiField)
+  if ((!renderNgsiField) || (metadataVector.size() > 0))
   {
-    jh.addRaw("metadata", metadataVector.toJson(orderedMetadata));
+      jh.addRaw("metadata", metadataVector.toJson(orderedMetadata));
```
Comment (fgalan) [inline src/lib/ngsi/ContextAttribute.cpp:1151]:
```suggestion
    jh.addRaw("metadata", metadataVector.toJson(orderedMetadata));
```

File: src/lib/ngsi/ContextAttribute.cpp [hunk 4]
```diff
@@ -1164,10 +1185,10 @@ std::string ContextAttribute::toJsonValue(void)
   }
   else if (valueType == orion::ValueTypeString)
   {
-    std::string out = "\"";
-    out += toJsonString(stringValue);
-    out += '"';
-    return out;
+    //std::string out = "\"";
+    //out += toJsonString(stringValue);
+    //out += '"';
```
Comment (fgalan) [inline src/lib/ngsi/ContextAttribute.cpp:1190]:
```suggestion
```

File: doc/manuals/orion-api.md [hunk 5]
```diff
@@ -2507,6 +2510,64 @@ will trigger a notification like this:
...
+    "B": {
+      "value": "${A/100}",
+      "type": "Calculated"
+    }
+  },
+  "attrs": [ "B" ]
+}
+```
+
+`B` could be set to the desired `0.04` (if `A` is evaluated before `B`) or to the underised `null` (if `B` is evaluated before `A`), randomly.
```
Comment (fgalan) [inline doc/manuals/orion-api.md:2536]:
```suggestion
in the resulting notification `B` could be set to the desired `0.04` (if `A` is evaluated before `B`) or to the underised `null` (if `B` is evaluated before `A`), randomly.
```

File: doc/manuals/orion-api.md [hunk 6]
```diff
@@ -2507,6 +2510,64 @@ will trigger a notification like this:
...
+
+In order to overcome this problem, the `evalPriority` metadata can be used to define the evaluation order. It works this way:
+
+* `evalPriority` metadata is a number from 1 (first evaluation) to 100000 (last evaluation)
+* Expressions are evaluated in incresing order of priority
+* In case of ties, Orion does not guarantee a particular evaluation order. Thus, expressions in the same priority level must be considered independent or bad things would happen
+* If no `evalPriority` is set, default 100000 is used
+* `evalPriority` only has a meaning in `notification.httpCustom.ngsi` in subscriptions. As metadata in regular entities (e.g. an entity created with `POST /v2/entities`) Orion doesn't implements any semantic.
```
Comment (fgalan) [inline doc/manuals/orion-api.md:2544]:
```suggestion
* `evalPriority` only has a meaning in `notification.httpCustom.ngsi` in subscriptions. As metadata in regular entities (e.g. an entity created with `POST /v2/entities`) Orion doesn't implements any semantic for it
```

**Review by mapedraza — APPROVED** [review]
LGTM

**Comment by fgalan** [comment]
@fisuda this PR does some additions to English documnentation. It would be great if you could sync the Japanese translation in a PR, please. Thanks in advance!

**Comment by fisuda** [comment]
I sent the PR #4591. Thanks.

**Pull request merged**